Support uptime command

jtai opened the pull request:
This change adds support for the `uptime` and `date` commands. Unfortunately, because the responses to these commands lack newlines, some workarounds are required. I'm pretty new to golang and am generally open to better ways to implement this.

MakotoE commented on api.go:
```diff
@@ -102,6 +103,40 @@ func (a *API) NumSlots() (int, error) {
+}
+
+// Date returns the current date and time
+func (a *API) Date() (time.Time, error) {
```
Although the wrapper should cover as much of the underlying API as possible, I hardly think anyone will need this. If they do, they should just use `Exec()`. I don't think any of the "Standard Commands" should be added. Please remove `Date()`.

jtai replied:
For context, the reason I'm interested in this library is to build a F@H prometheus exporter. Being new to Go, I based it on the memcached one, which exported a metric for the current server time. But I agree, I won't be using that metric in my dashboards, and it makes sense not to wrap the "Standard Commands". Will remove.

MakotoE replied:
I admit I have the "you don't need it" mindset, but now that you mentioned the use case I see a good reason to keep it. Sorry.

Then comes the question of "should we implement `add <number> <number>`"? My answer would probably be: send a PR.

jtai replied:
I'm also pretty minimalistic, definitely no need to apologize. I think the stance of not wrapping the "Standard Commands" makes sense, and I don't actually understand why the telnet API exposes a scripting language.

For now, why don't we not wrap `date`, but export the helper for the `exec` hack? That way I can easily implement support for it in the application code -- it's just one `time.Parse` call. If in the future, there is a good reason to support the rest of the Standard Commands in this library, I can always switch to the library version.

(As a side note, I do need the library to support the `uptime` command, because I need the special FAH duration parser which is not exported.)

MakotoE replied:
Sounds good!

MakotoE commented on api.go:
```diff
@@ -102,6 +103,40 @@ func (a *API) NumSlots() (int, error) {
+// Uptime returns FAH uptime
+func (a *API) Uptime() (time.Duration, error) {
+	// work around for response to uptime command not including a trailing newline
+	s, err := a.Exec("eval \"$(uptime)\\n\"")
```
I think there should be a helper function for `Exec()` for dealing with these commands which don't end in newline. It should `eval` the sub-command like what you did. (I call it a sub-command because it's probably used for inlining.)

At first I was thinking of doing this:

```diff
diff --git a/api.go b/api.go
index 198fbfd..801f3d6 100644
--- a/api.go
+++ b/api.go
@@ -252,9 +252,9 @@ func readMessage(r telnet.Reader) (string, error) {
 
 		buffer.WriteByte(b[0])
 
-		const endOfMessage = "\n> "
+		const endOfMessage = "> "
 		if strings.HasSuffix(buffer.String(), endOfMessage) {
-			return strings.TrimPrefix(strings.TrimSuffix(buffer.String(), "\n> "), "\n"), nil
+			return strings.TrimPrefix(strings.TrimSuffix(buffer.String(), "> "), "\n"), nil
 		}
 	}
 }

```

But then that might cause issues with embedded `"> "` in the middle of a message.

Instead of `"eval \"$(uptime)\\n\""`, you can use a string literal like this `` `eval "$(info)\n"` `` to make it easier to read.

jtai replied:
Yes, I came to the same conclusion regarding embedded `>` in the response, which is how I came to the `eval` hack.

A helper is a good idea. Any suggestions for its name?

MakotoE replied:
What about

```go
// ExecEval can execute commands which do not return a trailing newline.
func (a *API) ExecEval(command string) (string, error) {
```

jtai commented:
Ok I added `ExecEval` and rebased against master, and while I was at it I rebased away the implementation of `Date` to clean up the commit history a bit.

Thanks again for the detailed code review and discussion, and for open sourcing this API client!

MakotoE commented on api.go:
```diff
@@ -106,6 +117,21 @@ func (a *API) NumSlots() (int, error) {
+func (a *API) Uptime() (time.Duration, error) {
+	s, err := a.ExecEval("uptime")
+	if err != nil {
+		return time.Duration(0), err
```
This can be shortened to `return 0, err`.

jtai replied:
Updated

MakotoE commented on api.go:
```diff
@@ -106,6 +117,21 @@ func (a *API) NumSlots() (int, error) {
 	return n, unmarshalPyON(s, &n)
 }
 
+// Uptime returns FAH uptime
```
Sorry to be picky; by Go convention, function comments should end with a period.

MakotoE commented on api.go:
```diff
@@ -79,6 +80,16 @@ func (a *API) Exec(command string) (string, error) {
 
+// ExecEval executes commands which do not return a trailing newline.
+func (a *API) ExecEval(command string) (string, error) {
+	s, err := a.Exec(fmt.Sprintf(`eval "$(%s)\n"`, command))
```
Being picky again; `err` should be checked here, not at the end. Even though it's the same thing in this particular case, later on mistakes can happen if left like this.

jtai replied:
Sorry, I'm not entirely what you mean by checking here. Something like this?

```diff
diff --git a/api.go b/api.go
--- a/api.go
+++ b/api.go
@@ -83,6 +83,9 @@ func (a *API) Exec(command string) (string, error) {
 // ExecEval executes commands which do not return a trailing newline.
 func (a *API) ExecEval(command string) (string, error) {
        s, err := a.Exec(fmt.Sprintf(`eval "$(%s)\n"`, command))
+       if err != nil {
+               return "", err
+       }

        // when formatting with a newline, the response contains an extra trailing backslash
        s = strings.TrimSuffix(s, "\\")
```

MakotoE replied:
Yes exactly, and at the very end it can `return strings.TrimSuffix(s, "\\"), nil`.

jtai replied:
Done!

MakotoE commented:
I'd be interested to see your prometheus exporter; please let me know if you make it public!

jtai commented:
I just published it!

[jtai/foldingathome_exporter](https://github.com/jtai/foldingathome_exporter)

MakotoE commented:
Amazing, thanks!
